python/pyspark/ml/tuning.py

```diff
@@ -730,13 +733,40 @@ def _fit(self, dataset):
             train = datasets[i][0].cache()
 
             tasks = _parallelFitTasks(est, train, eva, validation, epm, collectSubModelsParam)
-            for j, metric, subModel in pool.imap_unordered(lambda f: f(), tasks):
-                metrics[j] += (metric / nFolds)
-                if collectSubModelsParam:
-                    subModels[i][j] = subModel
 
-            validation.unpersist()
-            train.unpersist()
+            sub_task_failed = False
+
+            @inheritable_thread_target
+            def run_task(task):
+                if sub_task_failed:
+                    raise RuntimeError("Terminate this task because one of other task failed.")
+                return task()
+
+            try:
+                for j, metric, subModel in pool.imap_unordered(run_task, tasks):
+                    metrics[j] += (metric / nFolds)
+                    if collectSubModelsParam:
+                        subModels[i][j] = subModel
+            except:
+                sub_task_failed = True
+                raise
+            finally:
+                if sub_task_failed:
+                    if is_pinned_thread_mode():
+                        try:
+                            time.sleep(1)
+                            sc = dataset._sc
+                            sc.cancelJobGroup(sc.getLocalProperty("spark.jobGroup.id"))
+                        except:
+                            pass
+                    else:
+                        warnings.warn("CrossValidator {} fit call failed but some spark jobs "
+                                      "may still running for unfinished trials. Enable pyspark "
```

Comment:
Hm, why is it inconsistent with `TrainValidationSplit`? Seems like `TrainValidationSplit` will always cancel but here only cancel when pinned thread mode is on.